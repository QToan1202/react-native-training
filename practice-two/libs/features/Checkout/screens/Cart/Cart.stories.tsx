import { ComponentMeta, ComponentStory } from '@storybook/react-native'

import { Cart } from '../index'

export default {
  title: 'screens/Checkout/Cart',
  component: Cart,
} as ComponentMeta<typeof Cart>

const Template: ComponentStory<typeof Cart> = () => <Cart />

export const Default = Template.bind({})
Default.args = {}
