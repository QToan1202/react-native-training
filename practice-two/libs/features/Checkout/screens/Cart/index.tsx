import { useCallback, useMemo } from 'react'
import { NativeStackScreenProps } from '@react-navigation/native-stack'
import { ScrollView, YStack, Spinner } from 'tamagui'
import { useShallow } from 'zustand/react/shallow'

import {
  Button,
  Heading,
  Paragraph,
  TabBar,
  ICart,
  useCartStore,
  useOrderStore,
  Feature,
} from '@practice-two/shared'
import { RootStackParamList } from 'libs/shared/navigation/Stack'
import { Address, CartItem, Price } from '../../components'

export type CartScreenProps = NativeStackScreenProps<RootStackParamList, 'Cart'>

const Cart = ({ navigation }: CartScreenProps) => {
  const [isHydrated, address] = useOrderStore(
    useShallow((state) => [state.isHydrated, state.address])
  )
  const [isHydratedCart, cart] = useCartStore(useShallow((state) => [state.isHydrated, state.cart]))
  // eslint-disable-next-line react-hooks/exhaustive-deps
  const handleChangeAddress = useCallback(() => navigation.navigate('AddAddress'), [])
  // eslint-disable-next-line react-hooks/exhaustive-deps
  const handleNavigateAddCard = useCallback(() => navigation.navigate('Payment'), [])
  const renderAddress: React.JSX.Element | null = useMemo(() => {
    if (!isHydrated) return null
    if (!address) return null

    return (
      <Address
        name={`Deliver to ${address.name}`}
        streetAddress={address.streetAddress}
        onPress={handleChangeAddress}
      />
    )
  }, [isHydrated, address, handleChangeAddress])
  const renderCartList: React.JSX.Element | React.JSX.Element[] = useMemo(() => {
    if (!isHydratedCart) return <Spinner size="large" color="$color.primary" />
    if (!cart.length)
      return (
        <YStack padding="$space.4">
          <Heading
            content="Cart empty!!!"
            color="$color.gray_50"
            fontWeight="$3"
            textAlign="center"
          />
          <Paragraph
            content="Shopping with us and add anything you desire of"
            color="$color.gray_50"
            textAlign="center"
          />
        </YStack>
      )

    return cart.map(({ id, img, name, price, discountPrice, quantity }: ICart) => (
      <CartItem
        key={id}
        image={{ uri: img }}
        name={name}
        price={price}
        discountPrice={discountPrice}
        quantity={quantity}
      />
    ))
  }, [cart, isHydratedCart])

  return (
    <Feature feat="checkout">
      <ScrollView flex={1} backgroundColor="$color.bg_layer">
        {renderAddress || (
          <Button title="+ add new address" variant="quaternary" onPress={handleChangeAddress} />
        )}
        <YStack marginTop="$space.2.5" marginBottom="$space.10">
          {renderCartList}
          <Button
            title="remove"
            variant="quaternary"
            color="$color.gray_50"
            onPress={handleChangeAddress}
          />
        </YStack>
        <Price data={cart} deliveryFee={1.5} />
      </ScrollView>
      <TabBar title="Continue to Payment" onPress={handleNavigateAddCard} />
    </Feature>
  )
}

export default Cart
