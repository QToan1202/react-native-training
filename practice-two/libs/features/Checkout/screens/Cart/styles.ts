import { StyleSheet } from 'react-native'

import { COLORS } from '@practice-two/shared'

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: COLORS.BG_LAYER,
  },
  btnTitle: {
    color: COLORS.GRAY_50,
    opacity: 0.5,
  },
  itemList: {
    marginTop: 10,
    marginBottom: 60,
  },
})

export default styles
