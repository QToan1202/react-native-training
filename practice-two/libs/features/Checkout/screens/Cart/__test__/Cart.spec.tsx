import { render } from '@testing-library/react-native'

import Cart from '../index'

describe('Testing Cart screen', () => {
  it('Render screen successfully', () => {
    const component = render(<Cart />).toJSON()

    expect(component).toMatchSnapshot()
  })
})
