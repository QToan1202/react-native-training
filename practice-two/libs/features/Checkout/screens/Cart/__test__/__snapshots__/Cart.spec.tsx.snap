// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Cart screen Render screen successfully 1`] = `
[
  <RCTScrollView
    contentContainerStyle={
      {
        "backgroundColor": "#f6f9ff",
        "flex": 1,
      }
    }
  >
    <View>
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "backgroundColor": "#fff",
            "borderColor": "#fff",
            "borderRadius": 0,
            "borderWidth": 2,
            "opacity": 1,
            "paddingVertical": 13,
          }
        }
      >
        <View
          style={
            {
              "alignItems": "center",
              "columnGap": 6,
              "display": "flex",
              "flexDirection": "row",
              "justifyContent": "center",
            }
          }
        >
          <Text
            style={
              [
                {
                  "color": "#fff",
                  "fontFamily": "Montserrat_400Regular",
                },
                {
                  "fontSize": 16,
                },
                [
                  {
                    "color": "#000",
                  },
                  {
                    "fontFamily": "Montserrat_500Medium",
                    "textAlign": "center",
                    "textTransform": "capitalize",
                  },
                  undefined,
                ],
              ]
            }
          >
            + add new address
          </Text>
        </View>
      </View>
      <View
        style={
          {
            "marginBottom": 60,
            "marginTop": 10,
          }
        }
      >
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={false}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "backgroundColor": "#fff",
              "opacity": 1,
              "paddingBottom": 12,
              "paddingLeft": 16,
              "paddingTop": 29,
            }
          }
        >
          <View
            style={
              {
                "flexDirection": "row",
                "gap": 15,
              }
            }
          >
            <Image
              source={
                {
                  "testUri": "../../../assets/cart/item.png",
                }
              }
              style={
                {
                  "borderRadius": 10,
                  "height": 102,
                  "width": 102,
                }
              }
            />
            <View
              style={
                {
                  "justifyContent": "center",
                  "rowGap": 12.5,
                }
              }
            >
              <Text
                style={
                  [
                    {
                      "color": "#fff",
                      "fontFamily": "Montserrat_400Regular",
                    },
                    {
                      "fontSize": 14,
                    },
                    {
                      "color": "#4f4f4f",
                      "fontFamily": "Montserrat_500Medium",
                      "textTransform": "capitalize",
                    },
                  ]
                }
              >
                coca cola
              </Text>
              <View
                style={
                  [
                    {
                      "alignItems": "center",
                      "columnGap": 4,
                      "flexDirection": "row",
                    },
                  ]
                }
              >
                <Text
                  numberOfLines={1}
                  style={
                    [
                      {
                        "color": "#fff",
                        "fontFamily": "Montserrat_400Regular",
                      },
                      {
                        "fontSize": 14,
                      },
                      {
                        "color": "#33907C",
                        "fontFamily": "Montserrat_700Bold",
                        "fontSize": 18,
                      },
                    ]
                  }
                >
                  $25
                </Text>
                <Text
                  numberOfLines={1}
                  style={
                    [
                      {
                        "color": "#fff",
                        "fontFamily": "Montserrat_400Regular",
                      },
                      {
                        "fontSize": 14,
                      },
                      [
                        {
                          "color": "#4f4f4f",
                          "fontFamily": "Montserrat_500Medium",
                          "textTransform": "capitalize",
                        },
                        {
                          "textDecorationLine": "line-through",
                        },
                      ],
                    ]
                  }
                >
                  $50
                </Text>
                <Text
                  numberOfLines={1}
                  style={
                    [
                      {
                        "color": "#fff",
                        "fontFamily": "Montserrat_400Regular",
                      },
                      {
                        "fontSize": 14,
                      },
                      {
                        "color": "#4f4f4f",
                        "fontFamily": "Montserrat_500Medium",
                        "textTransform": "capitalize",
                      },
                    ]
                  }
                >
                  50% off
                </Text>
              </View>
              <Text
                style={
                  [
                    {
                      "color": "#fff",
                      "fontFamily": "Montserrat_400Regular",
                    },
                    {
                      "fontSize": 14,
                    },
                    {
                      "color": "#4f4f4f",
                      "fontFamily": "Montserrat_500Medium",
                      "textTransform": "capitalize",
                    },
                  ]
                }
              >
                qty: 1
              </Text>
            </View>
          </View>
        </View>
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "backgroundColor": "#fff",
              "borderColor": "#fff",
              "borderRadius": 0,
              "borderWidth": 2,
              "opacity": 1,
              "paddingVertical": 13,
            }
          }
        >
          <View
            style={
              {
                "alignItems": "center",
                "columnGap": 6,
                "display": "flex",
                "flexDirection": "row",
                "justifyContent": "center",
              }
            }
          >
            <Text
              style={
                [
                  {
                    "color": "#fff",
                    "fontFamily": "Montserrat_400Regular",
                  },
                  {
                    "fontSize": 16,
                  },
                  [
                    {
                      "color": "#000",
                    },
                    {
                      "fontFamily": "Montserrat_500Medium",
                      "textAlign": "center",
                      "textTransform": "capitalize",
                    },
                    {
                      "color": "#4f4f4f",
                      "opacity": 0.5,
                    },
                  ],
                ]
              }
            >
              remove
            </Text>
          </View>
        </View>
      </View>
      <View
        style={
          [
            {
              "backgroundColor": "#fff",
              "paddingBottom": 15,
              "paddingHorizontal": 15,
            },
            undefined,
          ]
        }
      >
        <Text
          style={
            [
              {
                "color": "#fff",
                "fontFamily": "Montserrat_700Bold",
                "fontSize": 24,
              },
              false,
              [
                {
                  "color": "#000",
                  "fontFamily": "Montserrat_500Medium",
                  "lineHeight": 20,
                  "textTransform": "capitalize",
                },
                {
                  "fontFamily": "Montserrat_600SemiBold",
                  "fontSize": 18,
                  "paddingVertical": 14,
                },
              ],
            ]
          }
        >
          price details
        </Text>
        <View
          style={
            [
              {
                "alignItems": "center",
                "columnGap": 4,
                "flexDirection": "row",
              },
              {
                "justifyContent": "space-between",
              },
            ]
          }
        >
          <View
            style={
              {
                "rowGap": 10,
              }
            }
          >
            <Text
              style={
                [
                  {
                    "color": "#fff",
                    "fontFamily": "Montserrat_400Regular",
                  },
                  {
                    "fontSize": 14,
                  },
                  {
                    "color": "#000",
                    "fontFamily": "Montserrat_500Medium",
                    "lineHeight": 20,
                    "textTransform": "capitalize",
                  },
                ]
              }
            >
              price (1 item)
            </Text>
            <Text
              style={
                [
                  {
                    "color": "#fff",
                    "fontFamily": "Montserrat_400Regular",
                  },
                  {
                    "fontSize": 14,
                  },
                  {
                    "color": "#000",
                    "fontFamily": "Montserrat_500Medium",
                    "lineHeight": 20,
                    "textTransform": "capitalize",
                  },
                ]
              }
            >
              delivery fee
            </Text>
          </View>
          <View
            style={
              [
                {
                  "rowGap": 10,
                },
                {
                  "alignItems": "flex-end",
                },
              ]
            }
          >
            <Text
              style={
                [
                  {
                    "color": "#fff",
                    "fontFamily": "Montserrat_400Regular",
                  },
                  {
                    "fontSize": 14,
                  },
                  {
                    "color": "#000",
                    "fontFamily": "Montserrat_500Medium",
                    "lineHeight": 20,
                    "textTransform": "capitalize",
                  },
                ]
              }
            >
              $25
            </Text>
            <Text
              style={
                [
                  {
                    "color": "#fff",
                    "fontFamily": "Montserrat_400Regular",
                  },
                  {
                    "fontSize": 14,
                  },
                  {
                    "color": "#000",
                    "fontFamily": "Montserrat_500Medium",
                    "lineHeight": 20,
                    "textTransform": "capitalize",
                  },
                ]
              }
            >
              info
            </Text>
          </View>
        </View>
        <View
          style={
            {
              "borderBottomColor": "#0000001a",
              "borderBottomWidth": 0.5,
              "marginLeft": -14,
              "marginTop": 24,
              "width": 750,
            }
          }
        />
        <View
          style={
            [
              {
                "alignItems": "center",
                "columnGap": 4,
                "flexDirection": "row",
              },
              {
                "justifyContent": "space-between",
              },
            ]
          }
        >
          <Text
            style={
              [
                {
                  "color": "#fff",
                  "fontFamily": "Montserrat_700Bold",
                  "fontSize": 24,
                },
                false,
                [
                  {
                    "color": "#000",
                    "fontFamily": "Montserrat_500Medium",
                    "lineHeight": 20,
                    "textTransform": "capitalize",
                  },
                  {
                    "fontFamily": "Montserrat_600SemiBold",
                    "fontSize": 18,
                    "paddingVertical": 14,
                  },
                ],
              ]
            }
          >
            total amount
          </Text>
          <Text
            style={
              [
                {
                  "color": "#fff",
                  "fontFamily": "Montserrat_700Bold",
                  "fontSize": 24,
                },
                false,
                [
                  {
                    "color": "#000",
                    "fontFamily": "Montserrat_500Medium",
                    "lineHeight": 20,
                    "textTransform": "capitalize",
                  },
                  {
                    "fontFamily": "Montserrat_600SemiBold",
                    "fontSize": 18,
                    "paddingVertical": 14,
                  },
                  {
                    "fontFamily": "Montserrat_700Bold",
                  },
                ],
              ]
            }
          >
            $25
          </Text>
        </View>
      </View>
    </View>
  </RCTScrollView>,
  <View
    style={
      [
        {
          "paddingHorizontal": 16,
          "paddingVertical": 22,
        },
        {
          "backgroundColor": "#33907C",
          "rowGap": 24,
        },
        {
          "paddingTop": 0,
        },
        [
          {
            "backgroundColor": "#fff",
            "paddingHorizontal": 32,
          },
          {
            "shadowColor": "#000",
            "shadowOffset": {
              "height": 12,
              "width": 0,
            },
            "shadowOpacity": 0.6,
            "shadowRadius": 16,
          },
          {
            "paddingBottom": 12,
            "paddingTop": 12,
          },
          undefined,
        ],
      ]
    }
  >
    <View
      style={
        [
          {
            "alignItems": "center",
            "columnGap": 4,
            "flexDirection": "row",
          },
          {
            "justifyContent": "space-between",
          },
        ]
      }
    >
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "backgroundColor": "#33907C",
            "borderColor": "#33907C",
            "borderRadius": 50,
            "borderWidth": 2,
            "flex": 1,
            "opacity": 1,
            "paddingVertical": 13,
          }
        }
      >
        <View
          style={
            {
              "alignItems": "center",
              "columnGap": 6,
              "display": "flex",
              "flexDirection": "row",
              "justifyContent": "center",
            }
          }
        >
          <Text
            style={
              [
                {
                  "color": "#fff",
                  "fontFamily": "Montserrat_400Regular",
                },
                {
                  "fontSize": 16,
                },
                [
                  {
                    "color": "#fff",
                  },
                  {
                    "fontFamily": "Montserrat_500Medium",
                    "textAlign": "center",
                    "textTransform": "capitalize",
                  },
                  undefined,
                ],
              ]
            }
          >
            Continue to Payment
          </Text>
        </View>
      </View>
    </View>
  </View>,
]
`;
